### PR TITLE
feat: [PL-26565]: Adding a API to validate secret path in secret manager.

### PR DESCRIPTION
**Pull request by boopesh007**
## Describe your changes
Added the support to validate the secrets which are created using the secret reference approach. Once the validation is successful, the user can save the secret. Until today we check the validity during run time in NG.
image
image

Test cases tested in Local 
1. Valid path of a secret - output (Validation is successful)
2. Invalid path of a secret - output (Validation is unsuccessful)
3. Passing SecretManagerId of KMS instead of vault- SecretManager error 

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/43292)

### CODE REVIEW

File: 120-ng-manager/contracts/openapi/v1/openapi.yaml [hunk 1]
```diff
@@ -588,6 +588,28 @@ paths:
         - Account Secret
       security:
         - x-api-key: []
+    parameters: []
+  '/v1/secrets/validateSecretRef':
```
Comment (mankrit-singh) [inline 120-ng-manager/contracts/openapi/v1/openapi.yaml:592]: /v1/secrets/validate-secret-ref

File: 120-ng-manager/contracts/openapi/v1/openapi.yaml [hunk 2]
```diff
@@ -654,6 +676,33 @@ paths:
         - Org Secret
       security:
         - x-api-key: []
+  '/v1/orgs/{org}/secrets/validateSecretRef':
```
Comment (mankrit-singh) [inline 120-ng-manager/contracts/openapi/v1/openapi.yaml:679]: /v1/orgs/{org}/secrets/validate-secret-ref

File: 120-ng-manager/contracts/openapi/v1/openapi.yaml [hunk 3]
```diff
@@ -588,6 +588,28 @@ paths:
         - Account Secret
       security:
         - x-api-key: []
+    parameters: []
+  '/v1/secrets/validateSecretRef':
+    post:
+      summary: Validate secret reference
+      operationId: validate-account-secret-ref
+      description: Validates if the secret at the secretManager path can be referenced
+      parameters:
```
Comment (mankrit-singh) [inline 120-ng-manager/contracts/openapi/v1/openapi.yaml:597]: Why take account Identifier in Query Param? It should be from header, please refer to Harness-Account header in other apis

File: 120-ng-manager/contracts/openapi/v1/openapi.yaml [hunk 4]
```diff
@@ -654,6 +676,33 @@ paths:
...
         - x-api-key: []
+  '/v1/orgs/{org}/secrets/validateSecretRef':
+    post:
+      summary: Validate secret reference
+      operationId: validate-org-secret-ref
+      description: Validates if the secret at the secretManager path can be referenced
+      parameters:
+        - schema:
+            type: string
+          in: query
```
Comment (mankrit-singh) [inline 120-ng-manager/contracts/openapi/v1/openapi.yaml:687]: here also

File: 120-ng-manager/contracts/openapi/v1/openapi.yaml [hunk 5]
```diff
@@ -721,6 +770,39 @@ paths:
         - Project Secret
       security:
         - x-api-key: []
+  '/v1/orgs/{org}/projects/{project}/secrets/validateSecretRef':
```
Comment (mankrit-singh) [inline 120-ng-manager/contracts/openapi/v1/openapi.yaml:773]: /v1/orgs/{org}/projects/{project}/secrets/validate-secret-ref

File: 120-ng-manager/contracts/openapi/v1/openapi.yaml [hunk 6]
```diff
@@ -721,6 +770,39 @@ paths:
...
+    post:
+      summary: Validate secret reference
+      operationId: validate-project-secret-ref
+      description: Validates if the secret at the secretManager path can be referenced
+      parameters:
+        - schema:
+            type: string
+          in: query
```
Comment (mankrit-singh) [inline 120-ng-manager/contracts/openapi/v1/openapi.yaml:781]: here

File: 120-ng-manager/contracts/openapi/v1/openapi.yaml [hunk 7]
```diff
@@ -1659,6 +1741,30 @@ components:
           type: string
           description: Microservices version
       description: Microservice Version Info
+    SecretValidationMetadata:
+      title: SecretValidationMetadata
+      type: object
+      description: Details of the secret reference
+      properties:
+        secretManagerIdentifer:
```
Comment (mankrit-singh) [inline 120-ng-manager/contracts/openapi/v1/openapi.yaml:1749]: secret_manager_identifier

File: 120-ng-manager/contracts/openapi/v1/openapi.yaml [hunk 8]
```diff
@@ -1659,6 +1741,30 @@ components:
...
+          type: string
+          description: Identifier of the secretManager
+        secretRefPath:
```
Comment (mankrit-singh) [inline 120-ng-manager/contracts/openapi/v1/openapi.yaml:1752]: secret_ref_path

**Review by mankrit-singh — CHANGES_REQUESTED** [review]

File: 120-ng-manager/contracts/openapi/v1/openapi.yaml [hunk 9]
```diff
@@ -2075,6 +2181,12 @@ components:
           examples:
             sample:
               value: '<{url}?page=2&limit=1>; rel="next", <{url}?page=1&limit=1>; rel="self", <{url}?page=0&limit=1>; rel="previous", '
+    SecretValidationResponse:
```
Comment (mankrit-singh) [inline 120-ng-manager/contracts/openapi/v1/openapi.yaml:2184]: also, you can add an example in Stoplight, for Request also

**Review by mankrit-singh — COMMENTED** [review]
> Reviewed 5 of 10 files at r1, 5 of 5 files at r2, 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43292)* status: 11 file(s) to review. 1 change requests. 0 of 2 approvals obtained.

**Review by mankrit-singh — APPROVED** [review]
> Reviewed 7 of 7 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43292)* status: 11 file(s) to review. 1 of 2 approvals obtained.

**Review by nishantsaini20 — CHANGES_REQUESTED** [review]
> Reviewed 2 of 10 files at r1, 2 of 5 files at r2, 4 of 7 files at r4, 3 of 3 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43292)* status: 3 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`120-ng-manager/src/main/java/io/harness/ng/core/api/impl/NGEncryptedDataServiceImpl.java` line 561 at r5](https://reviewable.io/reviews/harness/harness-core/43292#-NNBxaho93sFYJs1S5dD:-NNBxaho93sFYJs1S5dE:b5g3l) ([raw file](https://github.com/harness/harness-core/blob/7b4f856532bdbde041e94d9a6f31479cb97f6129/120-ng-manager/src/main/java/io/harness/ng/core/api/impl/NGEncryptedDataServiceImpl.java#L561)):*
> ```Java
>   @Override
>   public boolean validateSecretRef(String accountIdentifier, String orgIdentifier, String projectIdentifier,
>       String secretManagerId, String secretRefPath) {
> ```

secretManagerIdentifier



_Code quote:_
```Java
secretManagerId
```
___
*[`120-ng-manager/src/main/java/io/harness/ng/core/api/impl/NGEncryptedDataServiceImpl.java` line 571 at r5](https://reviewable.io/reviews/harness/harness-core/43292#-NNByauz6sUUqb0UTegX:-NNByauz6sUUqb0UTegY:bg6ntgd) ([raw file](https://github.com/harness/harness-core/blob/7b4f856532bdbde041e94d9a6f31479cb97f6129/120-ng-manager/src/main/java/io/harness/ng/core/api/impl/NGEncryptedDataServiceImpl.java#L571)):*
> ```Java
>               .validateReference(accountIdentifier, secretRefPath, SecretManagerConfigMapper.fromDTO(secretManager));
>     } catch (Exception e) {
>       log.warn("Secret  reference validation failed for secretManager: {}", secretManager.getIdentifier(), e);
> ```

We should propagate the real error message to UI. The reason of failure can be wrong credentials of SM or no permission to read etc.
___
*[`120-ng-manager/src/main/java/io/harness/ng/core/remote/AccountSecretApiImpl.java` line 146 at r5](https://reviewable.io/reviews/harness/harness-core/43292#-NNBzA7f2d5GzQPqrpH6:-NNBzA7f2d5GzQPqrpH7:b2f7sf8) ([raw file](https://github.com/harness/harness-core/blob/7b4f856532bdbde041e94d9a6f31479cb97f6129/120-ng-manager/src/main/java/io/harness/ng/core/remote/AccountSecretApiImpl.java#L146)):*
> ```Java
>         account, null, null, body.getSecretManagerIdentifier(), body.getSecretRefPath());
>     if (!isValid) {
>       throw new NotFoundException(validationErrorMessage);
> ```

Follow up of previous comment
___
*[`120-ng-manager/src/main/java/io/harness/ng/core/remote/OrgSecretApiImpl.java` line 136 at r5](https://reviewable.io/reviews/harness/harness-core/43292#-NNBzIgu7efa01Hvxcmv:-NNBzIgu7efa01Hvxcmw:bb7od0r) ([raw file](https://github.com/harness/harness-core/blob/7b4f856532bdbde041e94d9a6f31479cb97f6129/120-ng-manager/src/main/java/io/harness/ng/core/remote/OrgSecretApiImpl.java#L136)):*
> ```Java
>     boolean isValid = ngEncryptedDataService.validateSecretRef(
>         account, org, null, body.getSecretManagerIdentifier(), body.getSecretRefPath());
>     if (!isValid) {
> ```

same
___
*[`120-ng-manager/src/main/java/io/harness/ng/core/remote/ProjectSecretApiImpl.java` line 141 at r5](https://reviewable.io/reviews/harness/harness-core/43292#-NNBzKBI2ypaJm64eO2m:-NNBzKBI2ypaJm64eO2n:bb7od0r) ([raw file](https://github.com/harness/harness-core/blob/7b4f856532bdbde041e94d9a6f31479cb97f6129/120-ng-manager/src/main/java/io/harness/ng/core/remote/ProjectSecretApiImpl.java#L141)):*
> ```Java
>     boolean isValid = ngEncryptedDataService.validateSecretRef(
>         account, org, project, body.getSecretManagerIdentifier(), body.getSecretRefPath());
>     if (!isValid) {
> ```

same
___
*[`120-ng-manager/src/test/java/io/harness/ng/core/remote/SecretApiImplTest.java` line 89 at r5](https://reviewable.io/reviews/harness/harness-core/43292#-NNBzVkQ7jNxoeP8_Hek:-NNBzVkQ7jNxoeP8_Hel:b-5lzboz) ([raw file](https://github.com/harness/harness-core/blob/7b4f856532bdbde041e94d9a6f31479cb97f6129/120-ng-manager/src/test/java/io/harness/ng/core/remote/SecretApiImplTest.java#L89)):*
> ```Java
>         ngSecretService, mock(SecretPermissionValidator.class), secretApiUtils, ngEncryptedDataService);
>     projectSecretApi = new ProjectSecretApiImpl(
>         ngSecretService, mock(SecretPermissionValidator.class), secretApiUtils, ngEncryptedDataService);
> ```

test case for validateProjectSecretRef and for other scopes as well

**Comment by boopesh007** [comment]
*[`120-ng-manager/src/test/java/io/harness/ng/core/remote/SecretApiImplTest.java` line 89 at r5](https://reviewable.io/reviews/harness/harness-core/43292#-NNBzVkQ7jNxoeP8_Hek:-NNCUJcwAIqyfqblLviw:b5v3htg) ([raw file](https://github.com/harness/harness-core/blob/7b4f856532bdbde041e94d9a6f31479cb97f6129/120-ng-manager/src/test/java/io/harness/ng/core/remote/SecretApiImplTest.java#L89)):*
<details><summary><i>Previously, nishantsaini20 (Nishant Saini) wrote…</i></summary><blockquote>

test case for validateProjectSecretRef and for other scopes as well
</blockquote></details>

all the APIs are calling the same function in the end. isnt testing the function layer sufficient.

**Review by nishantsaini20 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43292)* status: 3 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`120-ng-manager/src/test/java/io/harness/ng/core/remote/SecretApiImplTest.java` line 89 at r5](https://reviewable.io/reviews/harness/harness-core/43292#-NNBzVkQ7jNxoeP8_Hek:-NNCVTWP3fuK1FwR7hBA:bl2pun7) ([raw file](https://github.com/harness/harness-core/blob/7b4f856532bdbde041e94d9a6f31479cb97f6129/120-ng-manager/src/test/java/io/harness/ng/core/remote/SecretApiImplTest.java#L89)):*
<details><summary><i>Previously, boopesh007 (Boopesh Shanmugam) wrote…</i></summary><blockquote>

all the APIs are calling the same function in the end. isnt testing the function layer sufficient. 
</blockquote></details>

Both are units in themselves. If the caller start calling a different service layer method then it will be behavioural change for validateProjectSecretRef method which should be tested.

**Review by nishantsaini20 — COMMENTED** [review]
> Reviewed 6 of 6 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43292)* status: Some of the checks are too old. 9 file(s) to review. 1 change requests. 0 of 2 approvals obtained.

**Comment by boopesh007** [comment]
*[`120-ng-manager/src/main/java/io/harness/ng/core/api/impl/NGEncryptedDataServiceImpl.java` line 561 at r5](https://reviewable.io/reviews/harness/harness-core/43292#-NNBxaho93sFYJs1S5dD:-NN_K-FS83lLjST7w-kj:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/7b4f856532bdbde041e94d9a6f31479cb97f6129/120-ng-manager/src/main/java/io/harness/ng/core/api/impl/NGEncryptedDataServiceImpl.java#L561)):*
<details><summary><i>Previously, nishantsaini20 (Nishant Saini) wrote…</i></summary><blockquote>

secretManagerIdentifier
</blockquote></details>

Done.

**Comment by boopesh007** [comment]
*[`120-ng-manager/src/main/java/io/harness/ng/core/api/impl/NGEncryptedDataServiceImpl.java` line 571 at r5](https://reviewable.io/reviews/harness/harness-core/43292#-NNByauz6sUUqb0UTegX:-NN_K13BEZwX9fTpXeAo:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/7b4f856532bdbde041e94d9a6f31479cb97f6129/120-ng-manager/src/main/java/io/harness/ng/core/api/impl/NGEncryptedDataServiceImpl.java#L571)):*
<details><summary><i>Previously, nishantsaini20 (Nishant Saini) wrote…</i></summary><blockquote>

We should propagate the real error message to UI. The reason of failure can be wrong credentials of SM or no permission to read etc.
</blockquote></details>

Done.

**Comment by boopesh007** [comment]
*[`120-ng-manager/src/main/java/io/harness/ng/core/remote/AccountSecretApiImpl.java` line 146 at r5](https://reviewable.io/reviews/harness/harness-core/43292#-NNBzA7f2d5GzQPqrpH6:-NN_K2964tASlobPsk9J:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/7b4f856532bdbde041e94d9a6f31479cb97f6129/120-ng-manager/src/main/java/io/harness/ng/core/remote/AccountSecretApiImpl.java#L146)):*
<details><summary><i>Previously, nishantsaini20 (Nishant Saini) wrote…</i></summary><blockquote>

Follow up of previous comment
</blockquote></details>

Done.

**Comment by boopesh007** [comment]
*[`120-ng-manager/src/main/java/io/harness/ng/core/remote/OrgSecretApiImpl.java` line 136 at r5](https://reviewable.io/reviews/harness/harness-core/43292#-NNBzIgu7efa01Hvxcmv:-NN_K3Fw1xlA_NyrhQXk:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/7b4f856532bdbde041e94d9a6f31479cb97f6129/120-ng-manager/src/main/java/io/harness/ng/core/remote/OrgSecretApiImpl.java#L136)):*
<details><summary><i>Previously, nishantsaini20 (Nishant Saini) wrote…</i></summary><blockquote>

same
</blockquote></details>

Done.

**Comment by boopesh007** [comment]
*[`120-ng-manager/src/main/java/io/harness/ng/core/remote/ProjectSecretApiImpl.java` line 141 at r5](https://reviewable.io/reviews/harness/harness-core/43292#-NNBzKBI2ypaJm64eO2m:-NN_K4OW-XBDq7_DCarf:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/7b4f856532bdbde041e94d9a6f31479cb97f6129/120-ng-manager/src/main/java/io/harness/ng/core/remote/ProjectSecretApiImpl.java#L141)):*
<details><summary><i>Previously, nishantsaini20 (Nishant Saini) wrote…</i></summary><blockquote>

same
</blockquote></details>

Done.

**Comment by boopesh007** [comment]
*[`120-ng-manager/src/test/java/io/harness/ng/core/remote/SecretApiImplTest.java` line 89 at r5](https://reviewable.io/reviews/harness/harness-core/43292#-NNBzVkQ7jNxoeP8_Hek:-NN_K5IV7o2F_cDo34dx:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/7b4f856532bdbde041e94d9a6f31479cb97f6129/120-ng-manager/src/test/java/io/harness/ng/core/remote/SecretApiImplTest.java#L89)):*
<details><summary><i>Previously, nishantsaini20 (Nishant Saini) wrote…</i></summary><blockquote>

> caller
By caller here I mean validateProjectSecretRef (and other api layer methods)
</blockquote></details>

Done.

**Review by nishantsaini20 — APPROVED** [review]
> Reviewed 5 of 5 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43292)* status: 10 file(s) to review. 1 of 2 approvals obtained.

**Review by mankrit-singh — APPROVED** [review]
> Reviewed 3 of 3 files at r5, 1 of 6 files at r6, 5 of 5 files at r7, 1 of 1 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43292)* status: 1 file(s) to review. 1 of 2 approvals obtained.

**Review by shashank-harness — COMMENTED** [review]
Just do testing for KMS and handle that case.

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43292#-:-NNjeQd_BT-azCcdukQX:b-aa9r9k)* status: 1 file(s) to review. 1 of 2 approvals obtained.

**Review by shashank-harness — APPROVED** [review]
> Reviewed 1 of 10 files at r1, 3 of 3 files at r5, 1 of 6 files at r6, 4 of 5 files at r7, 1 of 1 files at r8, 1 of 1 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43292)* status: Some of the checks are too old. 1 file(s) to review. 1 of 2 approvals obtained.

**Review by shashank-harness — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43292)* status: Some of the checks are too old. 1 file(s) to review. 1 of 2 approvals obtained.

**Review by mankrit-singh — APPROVED** [review]
> Reviewed 1 of 1 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43292)* status: :shipit: complete! ✓

**Pull request merged**